relayer/src/chain/cosmos.rs

```diff
@@ -954,61 +955,88 @@ impl Chain for CosmosSdkChain {
         Ok(Sequence::from(response.next_sequence_receive))
     }
 
-    /// Queries the packet data for all packets with sequences included in the request.
-    /// Note - there is no way to format the query such that it asks for Tx-es with either
-    /// sequence (the query conditions can only be AND-ed)
-    /// There is a possibility to include "<=" and ">=" conditions but it doesn't work with
-    /// string attributes (sequence is emmitted as a string).
-    /// Therefore, here we perform one tx_search for each query. Alternatively, a single query
-    /// for all packets could be performed but it would return all packets ever sent.
+    /// This function queries transactions for events matching certain criteria.
+    /// 1. Client Update request - returns a vector with at most one update client event
+    /// 2. Packet event request - returns at most one packet event for each sequence specified
+    ///    in the request.
+    ///    Note - there is no way to format the packet query such that it asks for Tx-es with either
+    ///    sequence (the query conditions can only be AND-ed).
+    ///    There is a possibility to include "<=" and ">=" conditions but it doesn't work with
+    ///    string attributes (sequence is emmitted as a string).
+    ///    Therefore, for packets we perform one tx_search for each sequence.
+    ///    Alternatively, a single query for all packets could be performed but it would return all
+    ///    packets ever sent.
     fn query_txs(&self, request: QueryTxRequest) -> Result<Vec<IbcEvent>, Error> {
         crate::time!("query_txs");
 
         match request {
             QueryTxRequest::Packet(request) => {
-                crate::time!("query_txs");
+                crate::time!("query_txs: query packet events");
 
                 let mut result: Vec<IbcEvent> = vec![];
 
                 for seq in &request.sequences {
-                    // query all Tx-es that include events related to packet with given port, channel and sequence
+                    // query first (and only) Tx that includes the event specified in the query request
                     let response = self
                         .block_on(self.rpc_client.tx_search(
                             packet_query(&request, *seq),
                             false,
                             1,
-                            1,
+                            1, // get only the first Tx matching the query
                             Order::Ascending,
                         ))
-                        .unwrap(); // todo
-
-                    let mut events =
-                        packet_from_tx_search_response(self.id(), &request, *seq, response)
-                            .map_or(vec![], |v| vec![v]);
-
-                    result.append(&mut events);
+                        .map_err(|e| Kind::Grpc.context(e))?;
+
+                    assert!(
+                        response.txs.len() <= 1,
```

Comment:
Shouldn't this be `response.txs.len() != 1`? Or alternatively, if we allow more than 1 response: `response.txs.len() < 1`?

Reply:
In `tx_search()` above we ask for a single Tx. This is because for a packet event of a specified type, path ans sequence number there can be at most one Tx that includes such an event. So only `response.txs.len()` 0 or 1 are valid.

Reply:
My bad!